lib/_stream_readable.js

```diff
@@ -68,11 +68,21 @@ function ReadableState(options, stream) {
   // the point at which it stops calling _read() to fill the buffer
   // Note: 0 is a valid value, means "don't call _read preemptively ever"
   var hwm = options.highWaterMark;
-  var defaultHwm = this.objectMode ? 16 : 16 * 1024;
-  this.highWaterMark = (hwm || hwm === 0) ? hwm : defaultHwm;
+  if (hwm == null) {
+    var defaultHwm = this.objectMode ? 16 : 16 * 1024;
+    this.highWaterMark = defaultHwm;
```

Comment:
Sure, I wanted to, but I think I got an opposing comment when I did the same in one of the first commits of #12593. I cannot find it though, so I will go ahead and change it.